resource_elasticsearch_odfe_role.go

```diff
@@ -0,0 +1,304 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/olivere/elastic/uritemplates"
+
+	elastic7 "github.com/olivere/elastic/v7"
+)
+
+func resourceElasticsearchOdfeRole() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceElasticsearchOdfeRoleCreate,
+		Read:   resourceElasticsearchOdfeRoleRead,
+		Update: resourceElasticsearchOdfeRoleUpdate,
+		Delete: resourceElasticsearchOdfeRoleDelete,
+		Schema: map[string]*schema.Schema{
+			"role_name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"cluster_permissions": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+			"index_permissions": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"index_patterns": &schema.Schema{
+							Type:     schema.TypeSet,
+							Optional: true,
+							Elem: &schema.Schema{
+								Type: schema.TypeString,
+							},
+						},
+						"fls": &schema.Schema{
+							Type:     schema.TypeSet,
+							Optional: true,
+							Elem: &schema.Schema{
+								Type: schema.TypeString,
+							},
+						},
+						"masked_fields": &schema.Schema{
+							Type:     schema.TypeSet,
+							Optional: true,
+							Elem: &schema.Schema{
+								Type: schema.TypeString,
+							},
+						},
+						"allowed_actions": &schema.Schema{
+							Type:     schema.TypeSet,
+							Optional: true,
+							Elem: &schema.Schema{
+								Type: schema.TypeString,
+							},
+						},
+					},
+				},
+			},
+			"tenant_permissions": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"tenant_patterns": &schema.Schema{
+							Type:     schema.TypeSet,
+							Optional: true,
+							Elem: &schema.Schema{
+								Type: schema.TypeString,
+							},
+						},
+						"allowed_actions": &schema.Schema{
+							Type:     schema.TypeSet,
+							Optional: true,
+							Elem: &schema.Schema{
+								Type: schema.TypeString,
+							},
+						},
+					},
+				},
+			},
+			"description": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+		},
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+	}
+}
+
+func resourceElasticsearchOdfeRoleCreate(d *schema.ResourceData, m interface{}) error {
+	_, err := resourceElasticsearchPutOdfeRole(d, m)
+
+	if err != nil {
+		return err
+	}
+
+	name := d.Get("role_name").(string)
+	d.SetId(name)
+	return resourceElasticsearchOdfeRoleRead(d, m)
+}
+
+func resourceElasticsearchOdfeRoleRead(d *schema.ResourceData, m interface{}) error {
+	res, err := resourceElasticsearchGetOdfeRole(d.Id(), m)
```

Comment:
Here, I would rewrite like this:
```suggestion
	res, err := resourceElasticsearchGetOdfeRole(d.Id(), m)
	if err != nil {
		if elastic7.IsNotFound(err) {
			log.Printf("[WARN] OdfeRole (%s) not found, removing from state", d.Id())
			d.SetId("")
			return nil
		}
		return err
	}
```